### ISSUE TITLE
modifiers should generate branches checks on all  place they are used

### ISSUE THREAD

**Issue by wighawag** [sc-forks/solidity-coverage#618]
Currently `modifiers` are considered individually and not as part of where they are called
this cause coverage to wrongly assume the `modifier` code has been called on all branches simply because if it was executed in another context

solidity-coverage should copy-paste the `modifier` code and generate branching check on that resulting code.

For now, the only workaround is to not use `modifier` and instead copy the modifier code in place.

for example:
```solidity
function transferOwnership(address newOwner) external onlyOwner {
  emit OwnershipTransferred(owner, newOwner);
  owner = newOwner;
}

function setFee(uint256 fee) external onlyOwner {
  _fee = fee;
}

modifier onlyOwner() {
  require(msg.sender == owner, "NOT_AUTHORIZED");
  _;
}
```

will generate 100% coverage of branches even if only the following test are executed
- transferOwnership from owner
- setFee from owner
- setFee from non-owner

the missing test for "transferOwnership from non-owner" will not be seen by the coverage. 

On the contrary the following code will correctly trigger solidity-coverage to see less than 100% coverage as the `require(...` branching will be detected

```solidity
function transferOwnership(address newOwner) external {
  require(msg.sender == owner, "NOT_AUTHORIZED");
  emit OwnershipTransferred(owner, newOwner);
  owner = newOwner;
}
function setFee(uint256 fee) external  {
  require(msg.sender == owner, "NOT_AUTHORIZED");
  _fee = fee;
}
```

**Comment by cgewecke** [comment]
@wighawag Yes!

There is an 0.8.x beta which addresses this issue. See [this release note][1]. 

It was published at the beginning of January so there are some more recent bug fixes missing - I think it's equivalent to `0.7.13`.

Do you have any interest in trying it out? Would love some real-world feedback about whether or not things work.  


[1]: https://github.com/sc-forks/solidity-coverage/releases/tag/v0.8.0-beta.0

**Comment by wighawag** [comment]
Cool, I tried to find existing issue, but could not find (I think I was search for "modifiers" with a "s" :)
This is a duplicate of : https://github.com/sc-forks/solidity-coverage/issues/286
I ll close and I ll happily try the beta when I find some time. I ll report back
